// react-entities/lib/makeEntity.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = exports.makeEntity = exports.createEntity = exports.bindActions = exports.createSetState = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _store = require("./store");

var _useEntity = _interopRequireDefault(require("./useEntity"));

var createSetState = function createSetState(entity, beforeSetState) {
  return function (updates, updaterArg) {
    if (typeof updates === 'function') updates = updates(entity.state, updaterArg);
    if (typeof beforeSetState === 'function') beforeSetState(entity.state, updates);
    entity.state = (0, _extends2["default"])({}, entity.state, {}, updates);

    for (var i = 0; i < entity.subscribers.length; i++) {
      if (typeof entity.subscribers[i] === 'function') entity.subscribers[i](entity.state);
    } // Cleanup any nullified subscribers due to possible
    // component unmounts caused by this app state change


    for (var _i = 0; _i < entity.subscribers.length; _i++) {
      if (typeof entity.subscribers[_i] !== 'function') entity.subscribers.splice(_i, 1);
    }
  };
};

exports.createSetState = createSetState;

var bindActions = function bindActions(actions, entity, deps) {
  var entityActions = {};

  for (var key in actions) {
    if (typeof actions[key] === 'function') {
      var action = actions[key](entity, deps);
      if (typeof action !== 'function') throw new Error('Action must be defined using higher-order function.');
      entityActions[key] = action;
    }
  }

  return entityActions;
};

exports.bindActions = bindActions;

var createEntity = function createEntity(_ref, deps) {
  var initialState = _ref.initialState,
      _ref$options = _ref.options,
      options = _ref$options === void 0 ? {} : _ref$options,
      actions = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["initialState", "options"]);
  var id = (0, _store.reserveNextEntityId)();
  var entity = _store.store[id] = {
    state: initialState || {},
    initialState: initialState,
    subscribers: [],
    reset: function reset() {
      entity.state = initialState;
    }
  };
  entity.setState = createSetState(entity, options.beforeSetState);
  entity.actions = bindActions(actions, entity, deps);
  return entity;
};

exports.createEntity = createEntity;

var makeEntity = function makeEntity(definition, deps) {
  var entity = createEntity(definition, deps);
  return function (selector, equalityFn) {
    return (0, _useEntity["default"])(entity, selector, equalityFn);
  };
};

exports.makeEntity = makeEntity;
var _default = makeEntity;
exports["default"] = _default;
